### PR TITLE
[WIP] :seedling:  add e2e test for ownerRef reconciliation

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add a new e2e test and set of helper functions to test that ownerReferences are correctly reconciled by Cluster API.

This test is designed to capture how Cluster API creates ownerReferences with a newly created Topology based Cluster using KCP + KubeadmBooststrap + Docker infrastructure.

It also tests how ownerReferences are reconciled after they've been removed and re-reconciled.

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
Findings so far:

- Secrets are not re-reconciled with OwnerRefs #6979 
- KubeadmConfigTemplate are not re-reconciled with OwnerRefs when managed by a Cluster topology.
- KubeadmControlPlaneTemplate  are not re-reconciled with OwnerRefs when managed by a Cluster topology.
- A big diff in the way Docker infra resource ownerReferences are reconciled
- KubeadmControlPlane managed Machines are re-reconciled with both Cluster and KubeadmControlPlane owners (related to) #7529)
- KubeadmConfigs are created with two ownerRefs (one for Machine and one for MachineSet) but are only rereconciled with a Machine ownerRef

Errors from the most recent run:
```
  Unexpected ownerReferences for DockerMachineTemplate/quick-start-default-worker-machinetemplate: [wanted [cluster.x-k8s.io/v1beta1, Kind=Cluster], actual [], wanted [cluster.x-k8s.io/v1beta1, Kind=ClusterClass], actual []]
  Unexpected ownerReferences for KubeadmConfig/owner-references-rbl7zy-ptjnt-4rbmd: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-kubeconfig: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-ptjnt-4rbmd: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-md-0-bootstrap-77jnf-lpbng: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for Machine/owner-references-rbl7zy-ptjnt-7wxqt: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster]]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-etcd: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-ca: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-proxy: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for KubeadmConfigTemplate/quick-start-default-worker-bootstraptemplate: [wanted [cluster.x-k8s.io/v1beta1, Kind=Cluster], actual [], wanted [cluster.x-k8s.io/v1beta1, Kind=ClusterClass], actual []]
  Unexpected ownerReferences for KubeadmControlPlaneTemplate/quick-start-control-plane: wanted [cluster.x-k8s.io/v1beta1, Kind=ClusterClass], actual []
  Unexpected ownerReferences for DockerClusterTemplate/quick-start-cluster: wanted [cluster.x-k8s.io/v1beta1, Kind=ClusterClass], actual []
  Unexpected ownerReferences for KubeadmConfig/owner-references-rbl7zy-md-0-bootstrap-77jnf-lpbng: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Secret/owner-references-rbl7zy-sa: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for DockerMachine/owner-references-rbl7zy-control-plane-lw2tx-7lvkj: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for DockerMachineTemplate/quick-start-control-plane: [wanted [cluster.x-k8s.io/v1beta1, Kind=Cluster], actual [], wanted [cluster.x-k8s.io/v1beta1, Kind=ClusterClass], actual []]
  Unexpected ownerReferences for DockerMachine/owner-references-rbl7zy-md-0-infra-q9jqp-s8t2l: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
```

File: test/e2e/owner_references.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		}, clusterResources)
+
+		// Check that the references are as expected on cluster creation.
+		framework.AssertOwnerReferences(namespace.Name,
+			framework.CoreTypeOwnerReferenceAssertion,
+			framework.ExpOwnerReferenceAssertions,
+			framework.DockerInfraOwnerReferenceAssertions,
+			framework.KubeadmBootstrapOwnerReferenceAssertions,
+			framework.KubeadmControlPlaneOwnerReferenceAssertions,
```
Comment (fabriziopandini) [inline test/e2e/owner_references.go:95]: I think we should make it pluggable the list of provider-specific assertions, so the test could be reused by other providers as well

File: test/framework/ownerreference_helpers.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+	},
+	dockerClusterKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+	dockerClusterTemplateKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterClassGVK})
+	},
+}
+
+func hasExactOwnersByGVK(refList []metav1.OwnerReference, wantGVKs []schema.GroupVersionKind) error {
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:220]: What about
```suggestion
func hasOwners(refList []metav1.OwnerReference, gvk ...schema.GroupVersionKind) error {
```

File: test/framework/ownerreference_helpers.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	sort.SliceStable(wantGVKs, func(i int, j int) bool {
+		return wantGVKs[i].String() > wantGVKs[j].String()
+	})
+	if !reflect.DeepEqual(wantGVKs, refGVKs) {
+		return errors.New(fmt.Sprintf("wanted %v, actual %v", wantGVKs, refGVKs))
+	}
+	return nil
+}
+
+func hasOneOfExactOwnersByGVK(refList []metav1.OwnerReference, possibleGVKS ...[]schema.GroupVersionKind) error {
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:241]:
```suggestion
func hasOneOfOwners(refList []metav1.OwnerReference, possibleGVKS ...[]schema.GroupVersionKind) error {
```

File: test/framework/ownerreference_helpers.go (hunk 2 above)
Comment (fabriziopandini) [reply test/framework/ownerreference_helpers.go:220]: Also have you considered using sets.String given that at the end you are making the comparison on strings?

File: test/framework/ownerreference_helpers.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK, clusterResourceSetGVK})
+	},
+}
+
+var (
+	configMapKind = "ConfigMap"
+	secretKind    = "Secret"
+)
+
+var SecretOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{}
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:155]: Might be I'm wrong, but I was expecting Secrets to be owned by Cluster or ControlPlane (e.g CA, Kubeconfig) or by machines (Boostrap secrets)

File: test/framework/ownerreference_helpers.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+var ConfigMapReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	configMapKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterResourceSetGVK}, []schema.GroupVersionKind{})
+	},
+	secretKind: func(owners []metav1.OwnerReference) error {
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:160]: This being nested in ConfigMapReferenceAssertions slightly confuses me... might be I'm missing something

File: test/framework/ownerreference_helpers.go (hunk 4 above)
Comment (fabriziopandini) [reply test/framework/ownerreference_helpers.go:155]: Also, we should provide comments about what each assertion is testing

File: test/framework/ownerreference_helpers.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+var (
+	kubeadmConfigKind         = "KubeadmConfig"
+	kubeadmConfigTemplateKind = "KubeadmConfigTemplate"
+
+	kubeadmConfigGVK = bootstrapv1.GroupVersion.WithKind(kubeadmConfigKind)
+)
+
+var KubeadmBootstrapOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	kubeadmConfigTemplateKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK}, []schema.GroupVersionKind{clusterClassGVK})
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:191]: I think I understand why we have two options here (kubeadmConfigTemplate in a ClusterClass or in Cluster), but I'm wondering if there is a way that we can use to make more specific assertions, might be using the concept of tenant that exists in clusterctl discovery...
Same below for the dockerMachineTemplate or for everything were we use hasOneOfExactOwnersByGVK

File: test/framework/ownerreference_helpers.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+	},
+	kubeadmConfigKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{machineGVK, machineSetGVK}, []schema.GroupVersionKind{machineGVK, kubeadmControlPlaneGVK})
```
Comment (fabriziopandini) [inline test/framework/ownerreference_helpers.go:194]: is the kubeadmConfigKind owned by a MachineSet? I have to double-check why this...
Same below for Docker machine owned by MachineSet/KCP

**Review by fabriziopandini — COMMENTED** [review]
@killianmuldoon thanks for writing this test, this work is discovering really interesting stuff that will help to improve CAPI robustness.

Overall I really like the approach, which is simple even if tackling a complex problem; I think we can work a little bit on assertion functions making them a little bit more specific, but TBD if now or in a follow up iteration

WRT to how to address the finding of the first tests, I will follow up next week

**Comment by killianmuldoon** [comment]
The current version on this branch (with some first effort fix commits) leaves the following errors on a Topology managed Cluster:

```
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-control-plane-ctrv6: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-control-plane-trgtn: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Secret/owner-references-yi29pw-kubeconfig: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-md-0-6hzrr: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Machine/owner-references-yi29pw-control-plane-w6tcn: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster]]
  Unexpected ownerReferences for Machine/owner-references-yi29pw-control-plane-6wcjv: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster]]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-control-plane-9zs25: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-control-plane-9q8jx: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-md-0-c4kcs: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
```

And for a non-ClusterClass Cluster:
```
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-control-plane-ctrv6: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-control-plane-trgtn: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Secret/owner-references-yi29pw-kubeconfig: [wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [], wanted [bootstrap.cluster.x-k8s.io/v1beta1, Kind=KubeadmConfig], actual []]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-md-0-6hzrr: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for Machine/owner-references-yi29pw-control-plane-w6tcn: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster]]
  Unexpected ownerReferences for Machine/owner-references-yi29pw-control-plane-6wcjv: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane], actual [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Cluster]]
  Unexpected ownerReferences for KubeadmConfig/owner-references-yi29pw-control-plane-9zs25: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-control-plane-9q8jx: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]
  Unexpected ownerReferences for DockerMachine/owner-references-yi29pw-md-0-c4kcs: [wanted [cluster.x-k8s.io/v1beta1, Kind=MachineSet cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine], wanted [controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane cluster.x-k8s.io/v1beta1, Kind=Machine], actual [cluster.x-k8s.io/v1beta1, Kind=Machine]]

```

**Comment by killianmuldoon** [comment]
/close

Moving discussion to #7606 which has the actual proposed implementation.

**Pull request closed without merging**